Fix 188: Autocomplete for imports and triple slash reference paths

riknoll opened the pull request:
Fixes #188 
This adds text completions in imports and triple slash references for the following items:

For triple slash references:
* Relative and absolute paths

For imports and require:
* Relative and absolute paths
* Modules discovered by the type checker (via triple slash references, `@types` folder, `tsconfig.json`, etc.)
* Modules listed in your project's `package.json`

This is a breaking change that requires surfacing some of the `sys` object's file system reading APIs to the `LanguageServiceHost`

Note that this will only pick up node modules that are in `package.json`. This was done to improve efficiency and to make sure that sub-dependencies don't crowd the completion list.

msftclas commented:
Hi __@riknoll__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Richard Knoll). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

riknoll commented:
@paulvanbrenk please take a look when you have a moment

paulvanbrenk commented on src/compiler/checker.ts:
```diff
@@ -2,6 +2,8 @@
 
 /* @internal */
 namespace ts {
+    const ambientModuleSymbolRegex = /^".+"$/;
```
Can't this also be /^'.+'$/ ?

riknoll replied:
I was copying the behavior where we access the modules elsewhere in the checker (see `resolveExternalModuleNameWorker()`)

riknoll commented:
@vladima would you mind taking a look at this when you have a chance?

paulvanbrenk commented on src/harness/harnessLanguageService.ts:
```diff
@@ -135,7 +135,8 @@ namespace Harness.LanguageService {
         public getFilenames(): string[] {
             const fileNames: string[] = [];
-            ts.forEachValue(this.fileNameToScript, (scriptInfo) => {
+            this.virtualFileSystem.getAllFileEntries().forEach((virtualEntry) => {
```
Use ts.forEach.. since this function won't work downlevel

paulvanbrenk commented on src/harness/virtualFileSystem.ts:
```diff
@@ -157,9 +157,34 @@ namespace Utils {
             return directory;
         }
+
+        getAccessibleFileSystemEntries(path: string) {
```
Can you add a comment here, not sure what this method does

riknoll replied:
Done

paulvanbrenk commented on src/harness/virtualFileSystem.ts:
```diff
@@ -157,9 +157,34 @@ namespace Utils {
+            return fileEntries;
+
+            function getFilesRecursive(dir: VirtualDirectory<T>, result: VirtualFile<T>[]) {
+                dir.getFiles().forEach((e) => result.push(e));
```
See before about the foreach function, use ts.foreach

riknoll replied:
Thanks! I removed it everywhere in the PR

paulvanbrenk commented on src/services/services.ts:
```diff
@@ -1943,6 +1948,8 @@ namespace ts {
 
 
+    const tripleSlashDirectivePrefixRegex = /^\/\/\/\s*</;
+    const tripleSlashDirectiveFragmentRegex = /^\/\/\/\s*<reference\s+path\s*=\s*(?:'|")([^'"]+)$/;
```
Can you comments here, regexp are very hard to read/reason about

riknoll replied:
Fixed

riknoll commented:
@mhegazy or @RyanCavanaugh, mind taking a look?

mhegazy commented on src/services/services.ts:
```diff
@@ -2808,6 +2826,15 @@ namespace ts {
         }
     }
 
+    function isInReferenceComment(sourceFile: SourceFile, position: number): boolean {
```
probably should move this to utilities.

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+
+                return {
+                    isMemberCompletion: false,
+                    isNewIdentifierLocation: false,
```
this should be true to allow for the user defining a new module name that can not be found by us, e.g. `"file.json!json"` or `"~/folder"`

riknoll replied:
That's a good point. Made the change!

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+                    forEach(files, (f) => {
+                        const fName = includeExtensions ? getBaseFileName(f) : removeFileExtension(getBaseFileName(f));
+
+                        if (startsWith(fName, toComplete)) {
```
this means in VS you need to get completions (ctrl+space) after a backspace if the completion list is visible?

mhegazy replied:
i do not think we should filter here. the caller should.

riknoll replied:
That sounds good to me. I'm looking into that now.

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+                                name: fName,
+                                kind: ScriptElementKind.unknown,
+                                kindModifiers: ScriptElementKindModifier.none,
+                                sortText: fName
```
we will need to add a new concept of completion span (similar to Roslyn's, see: http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Completion/CompletionItem.cs,39).

The assumption now is the at the caller will compute the span as the first non-identifier character from the completion location. this is not true with file path, so we should provide a span to the caller.

the span should be the span of the text in the literal, moduluo the quotes. e.g. `".\mod"` => span of `.\mod`

mhegazy replied:
so just to be clear, we should add a new optional `span` member to `CompletionEntry`. the caller would use this to compute the replacement span.

paulvanbrenk replied:
The Roslyn API changed here for VS "15", there is no longer a span per item, but a single one for the completionlist.

mhegazy replied:
oh, i see now, you are only returning the file name and not the full module name.

nevertheless, the module name can contain non-identifier characters, e.g. `.`. we should compute the replacement span and send it back to the caller.

riknoll replied:
The replacement span is now sent back as part of the API refactor I did.

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+                        if (startsWith(fName, toComplete)) {
+                            result.push({
+                                name: fName,
+                                kind: ScriptElementKind.unknown,
```
we should add a new entry to SriptElmentKind for file/externalModule

riknoll replied:
Done

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+                            if (startsWith(dName, toComplete)) {
+                                result.push({
+                                    name: ensureTrailingDirectorySeparator(dName),
+                                    kind: ScriptElementKind.unknown,
```
we should add a new entry for `ScriptElementKind.Directory`

riknoll replied:
Done

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+                return ts.map(enumeratePotentialNonRelativeModules(fragment, scriptPath), (moduleName) => {
+                    return {
+                        name: moduleName,
+                        kind: ScriptElementKind.unknown,
```
`ScriptElementKind.ExternalModule`

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+            }
+
+            function enumeratePotentialNonRelativeModules(fragment: string, scriptPath: string) {
+                const ambientSymbolNameRegex = /^"(.+?)"$/;
```
lift this outside the function to avoid creating a new object every time the function is called.

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+
+                // Get modules that the type checker picked up
+                const ambientModules = ts.map(program.getTypeChecker().getAmbientModules(), (sym) => {
+                    const match = ambientSymbolNameRegex.exec(sym.name);
```
why do you need this? the checker should only return ambient modules.

riknoll replied:
You are correct, I've removed the check and the regex altogether

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+                // Complete the path by looking for source files and directories
+                const result: CompletionEntry[] = [];
+
+                const toComplete = getBaseFileName(fragment);
```
we need to handle the case of requesting completion on a file with an extension... e.g.

```ts
import d from ".\foo.js| // we should complete this to `./foo.js` and not just `./foo`
```

riknoll replied:
In general, should we prefer completions with file extensions or without?

In other words, is the desired behavior something like this:
```typescript
// At first, completion list given entries with no extension
import d from ".\fo";         // Completes to .\foo

// As soon as we see a ".", switch to including the extension
import d from ".\foo.";       // Completes to .\foo.ts
```
If so, that's fine with me. That distinction should probably be made by the caller if they are the ones doing the filtering. This code should just always return results with extensions.

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+                        return match[1];
+                    }
+                    // This should never happen
+                    return sym.name;
```
if `options.baseUrl` is set, you should include the top level directories of the baseURL. @vladima can explain how this works on the resolution side.

mhegazy replied:
also include `@types` pacakges, these are `options.types` || `readDirectory(options.typeRoots)` || automatically included type pacakges (i.e. contents of `.\node_modules\tsconfig.json`).

@RyanCavanaugh can talk to how this works in the compiler.

mhegazy replied:
also we need to add completions for `/// <reference types="..." ` based on the same logic.

riknoll replied:
Added logic for each of those. Are there any other compiler options I should be looking out for? The [documentation](https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/Module%20Resolution.md) doesn't mention any.

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+                    });
+
+                    // If possible, get folder completion as well
+                    if (host.getDirectories) {
```
if `options.rootDirs` is specified, we should enumerate other directories mapped to this directory. @vladima can explain how this works on the module resolution side.

riknoll replied:
Done! See the poorly named `getCompletionEntriesForDirectoryFragmentWithRootDirs()` above

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+
+                const toComplete = getBaseFileName(fragment);
+                const absolutePath = normalizeSlashes(host.resolvePath(isRootedDiskPath(fragment) ? fragment : combinePaths(scriptPath, fragment)));
+                const baseDir = toComplete ? getDirectoryPath(absolutePath) : absolutePath;
```
nit. please spell out `baseDirectory`

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+                if (directoryProbablyExists(baseDir, host)) {
+                    // Enumerate the available files
+                    const files = host.readDirectory(baseDir, extensions, /*exclude*/undefined, /*include*/["./*"]);
+                    forEach(files, (f) => {
```
nit: no parens

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+                    // Enumerate the available files
+                    const files = host.readDirectory(baseDir, extensions, /*exclude*/undefined, /*include*/["./*"]);
+                    forEach(files, (f) => {
+                        const fName = includeExtensions ? getBaseFileName(f) : removeFileExtension(getBaseFileName(f));
```
nit: fileName

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+                    if (host.getDirectories) {
+                        const directories = host.getDirectories(baseDir);
+                        forEach(directories, (d) => {
+                            const dName = getBaseFileName(removeTrailingDirectorySeparator(d));
```
nit: `directoryName`

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+                    }
+                }
+
+                return includeExtensions ? result : deduplicate(result, (a, b) => a.name === b.name);
```
dedupe before creating the entries. less garbage this way.

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+                const ambientSymbolNameRegex = /^"(.+?)"$/;
+
+                // If this is a nested module, get the module name
+                const firstSeparator = fragment.indexOf(directorySeparator);
```
we should normalize the name before checking for directory separator.

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+                // Nested modules of the form "module-name/sub" need to be adjusted to only return the string
+                // after the last '/' that appears in the fragment because editors insert the completion
+                // only after that character
+                nonRelativeModules = ts.map(nonRelativeModules, (moduleName) => {
```
make sure we normalize the path before comparing.

mhegazy replied:
nit: no parens

mhegazy commented on src/services/utilities.ts:
```diff
@@ -927,4 +933,101 @@ namespace ts {
         return ensureScriptKind(fileName, scriptKind);
     }
+
+    export function enumerateNodeModulesVisibleToScript(host: LanguageServiceHost, scriptPath: string, modulePrefix?: string) {
```
why is this in utilities? looks like the only user is in services.ts

mhegazy replied:
WOW.. this is a lot of IO just to get a completion entry. we do not have that much budget. if each package takes 1ms, and you have 200 packages in your node_modules, this is 200ms, we still need to create a checker, bind the whole program...

we should talk about this part. we should look into possible caching opportunities given that the package.json file is likely not change between invocations.

i would recommend not including this in the first iteration.

mhegazy commented on src/services/utilities.ts:
```diff
@@ -927,4 +933,101 @@ namespace ts {
+         *     The module has a "main" export and an index.d.ts file
+         *     The module has an index.ts
+         */
+        function moduleCanBeImported(modulePath: string): boolean {
```
does this even matter? why would you have a package in your node_modules that does not have any of these, and then why would you include it.

riknoll replied:
Some modules can't be required by their name alone, but instead define a bunch of submodules. For example, the "angular2" package cannot be required just the package name:
```Typescript
import * as a from "angular2/core";     // Works fine
import * as b from "angular2";          // Error
```
In that case, it's more helpful to return a completion with a trailing slash (e.g. "angular2/") or just give completions for all of the nested names. If this check is too expensive, I'm fine with removing it since that really is a minor thing.

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+                    return moduleName;
+                });
+
+                forEach(enumerateNodeModulesVisibleToScript(host, scriptPath, moduleNameFragment), visibleModule => {
```
condition this on `options.moduleResolution`

riknoll replied:
Do we plan on supporting completions for classic module resolution as well? I don't currently handle that in the code and that would likely be a pretty expensive check.

mhegazy commented on src/services/services.ts:
```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+                        nonRelativeModules.push(visibleModule.canBeImported ? visibleModule.moduleName : ensureTrailingDirectorySeparator(visibleModule.moduleName));
+                    }
+                    else {
+                        const nestedFiles = host.readDirectory(visibleModule.moduleDir, supportedTypeScriptExtensions, /*exclude*/undefined, /*include*/["./*"]);
```
this is a lot of IO again.. not sure we need to do all this work, specially that we are not caching any thing.

riknoll commented:
After the review by @mhegazy, I've done some major refactoring in addition to addressing feedback. Import completions now have their own API separate from other completions and are returned with an accompanying text span that indicates the text to be replaced. Additionally, completions are no longer filtered by the language service and should be filtered by the caller instead.

rbuckton commented on src/compiler/checker.ts:
```diff
@@ -19885,5 +19888,15 @@ namespace ts {
+        function getAmbientModules(): Symbol[] {
+            const result: Symbol[] = [];
+            for (const sym in globals) {
+                if (hasProperty(globals, sym) && ambientModuleSymbolRegex.test(sym)) {
```
You may want to update this branch from master to take advantage of a recent change around `Map<T>` performance. Anything declared `Map<T>` (such as the symbol table) now has a `null` prototype, so it is no longer necessary to check `hasProperty`.

rbuckton commented on src/compiler/program.ts:
```diff
@@ -10,9 +10,9 @@ namespace ts {
     const defaultTypeRoots = ["node_modules/@types"];
 
-    export function findConfigFile(searchPath: string, fileExists: (fileName: string) => boolean): string {
+    export function findConfigFile(searchPath: string, fileExists: (fileName: string) => boolean, configName="tsconfig.json"): string {
```
nit: spacing around `=` symbol.

rbuckton commented on src/harness/harness.ts:
```diff
@@ -458,6 +458,9 @@ namespace Harness {
     // harness always uses one kind of new line
     const harnessNewLine = "\r\n";
 
+    // Roote for file paths that are stored in a virtual file system
```
*Root

rbuckton commented on src/harness/virtualFileSystem.ts:
```diff
@@ -1,11 +1,11 @@
 namespace Utils {
-    export class VirtualFileSystemEntry {
-        fileSystem: VirtualFileSystem;
+    export class VirtualFileSystemEntry<T> {
```
Arbitrary types for a "file system" seems a bit odd to me.

rbuckton commented on src/harness/virtualFileSystem.ts:
```diff
@@ -157,9 +159,51 @@ namespace Utils {
+            }
+        }
+
+        normalizePathRoot(path: string) {
```
What is this doing, and why?

riknoll replied:
Whoops, that was from an old version of the code and I forgot to take it out. I'll remove it.

rbuckton commented on src/services/services.ts:
```diff
@@ -1166,6 +1170,11 @@ namespace ts {
         resolveTypeReferenceDirectives?(typeDirectiveNames: string[], containingFile: string): ResolvedTypeReferenceDirective[];
         directoryExists?(directoryName: string): boolean;
+
+        /**
```
This seems less like a jsdoc comment more of a dev note.

rbuckton commented on src/services/services.ts:
```diff
@@ -2034,7 +2064,13 @@ namespace ts {
+     * for completions.
+     * For example, this matches /// <reference path="fragment
+     */
+    const tripleSlashDirectiveFragmentRegex = /^(\/\/\/\s*<reference\s+(path|types)\s*=\s*(?:'|"))([^'"]*)$/;
```
The `$` token in the pattern will fail to match `/// <reference types="node" />` because a `"` is encountered before the end of the string/line.
The `'` character is valid (though rarely used) in a path, and should only be excluded if the attribute opens with `'`. 

Consider the following instead:

```js
/^(\/\/\/\s*<reference\s+(path|types)\s*=\s*('|"))(.*?)\3/
```

It pushes the match for the attribute value to index 4 when used below, but matches the quote characters and does not contain the `$` assertion.

riknoll replied:
The regex is actually only executed on the fragment of text leading up to the current caret position, so I believe matching the end of text is the correct behavior. I'll try and clarify that in the comment (I had the same thought earlier when I was reviewing my own code). Good catch on the `'` character, I'll make the change!

rbuckton commented on src/services/services.ts:
```diff
@@ -4459,6 +4497,466 @@ namespace ts {
+             * containing folder via the "rootDirs" compiler option
+             */
+            function getBaseDirectoriesFromRootDirs(rootDirs: string[], basePath: string, scriptPath: string, ignoreCase: boolean): string[] {
+                // Make all paths absolute/normalized if they are not already
```
How expensive is this to do for each keypress in a project like Angular? Do we need to employ some type of caching mechanism rather than recalculate this each time?

riknoll replied:
I did some quick measurements to get an idea about whether this is too much IO. I generated a project with 200 node modules and 150 @ types modules and timed the completions for non-relative (e.g. "some-module-name") and relative (e.g. "../node-modules/") paths. In general, the time for completions was always sub 200 ms. For more reasonably sized directories, the completions are more or less instantaneous. There are rare spikes (up to about 600-700 ms) and I think those are likely due to accessing storage that has fallen out of memory or when the hard drive is in a cold state. Those are pretty rare, however, and after a spike happens you don't see another for a while. For the record, I was timing end-to-end (not just the IO).

Completions shouldn't be triggered on each keypress, just after trigger characters like "/". The caller should handle the filtering of the results until the next trigger character.

rbuckton commented on src/services/services.ts:
```diff
@@ -4464,6 +4503,466 @@ namespace ts {
+
+                // Determine the path to the directory containing the script relative to the root directory it is contained within
+                let relativeDirectory: string;
+                forEach(rootDirs, rootDirectory => {
```
consider `for..of` to avoid the closure.

rbuckton commented on src/services/services.ts:
```diff
@@ -4464,6 +4503,466 @@ namespace ts {
+                if (directoryProbablyExists(baseDirectory, host)) {
+                    // Enumerate the available files
+                    const files = host.readDirectory(baseDirectory, extensions, /*exclude*/undefined, /*include*/["./*"]);
+                    forEach(files, filePath => {
```
consider `for..of` to avoid the closure.

rbuckton commented on src/services/services.ts:
```diff
@@ -4464,6 +4503,466 @@ namespace ts {
+                    // If possible, get folder completion as well
+                    if (host.getDirectories) {
+                        const directories = host.getDirectories(baseDirectory);
+                        forEach(directories, d => {
```
consider `for..of` to avoid the closure.

rbuckton commented on src/services/services.ts:
```diff
@@ -4464,6 +4503,466 @@ namespace ts {
+                            if (paths.hasOwnProperty(path)) {
+                                if (path === "*") {
+                                    if (paths[path]) {
+                                        forEach(paths[path], pattern => {
```
consider `for..of` to avoid the closure.

rbuckton commented on src/services/services.ts:
```diff
@@ -4464,6 +4503,466 @@ namespace ts {
+                                if (path === "*") {
+                                    if (paths[path]) {
+                                        forEach(paths[path], pattern => {
+                                            forEach(getModulesForPathsPattern(fragment, baseUrl, pattern, fileExtensions), match => {
```
consider `for..of` to avoid the closure.

rbuckton commented on src/services/services.ts:
```diff
@@ -4464,6 +4503,466 @@ namespace ts {
+
+                getCompletionEntriesFromTypings(host, options, scriptPath, result);
+
+                forEach(enumeratePotentialNonRelativeModules(fragment, scriptPath, options), moduleName => {
```
consider `for..of` to avoid the closure.

rbuckton commented on src/services/services.ts:
```diff
@@ -4464,6 +4503,466 @@ namespace ts {
+                    const result: string[] = [];
+
+                    // Trim away prefix and suffix
+                    forEach(matches, match => {
```
consider `for..of` to avoid the closure.

rbuckton commented on src/services/services.ts:
```diff
@@ -4464,6 +4503,466 @@ namespace ts {
+                const moduleNameFragment = isNestedModule ? fragment.substr(0, fragment.lastIndexOf(directorySeparator)) : undefined;
+
+                // Get modules that the type checker picked up
+                const ambientModules = ts.map(program.getTypeChecker().getAmbientModules(), sym => stripQuotes(sym.name));
```
unecessary `ts.` prefix

rbuckton commented on src/services/services.ts:
```diff
@@ -4464,6 +4503,466 @@ namespace ts {
+
+                // Get modules that the type checker picked up
+                const ambientModules = ts.map(program.getTypeChecker().getAmbientModules(), sym => stripQuotes(sym.name));
+                let nonRelativeModules = ts.filter(ambientModules, moduleName => startsWith(moduleName, fragment));
```
unnecessary `ts.` prefix.

rbuckton commented on src/services/services.ts:
```diff
@@ -4464,6 +4503,466 @@ namespace ts {
+                // starts
+                if (isNestedModule) {
+                    const moduleNameWithSeperator = ensureTrailingDirectorySeparator(moduleNameFragment);
+                    nonRelativeModules = ts.map(nonRelativeModules, moduleName => {
```
unnecessary `ts.` prefix here as well

riknoll commented:
I've added guards to the new language service host functionality so this should no longer be a breaking change.

rbuckton commented on src/services/services.ts:
```diff
@@ -4464,6 +4508,475 @@ namespace ts {
+            function getCompletionEntriesFromTypings(host: LanguageServiceHost, options: CompilerOptions, scriptPath: string, result: ImportCompletionEntry[] = []): ImportCompletionEntry[] {
+                // Check for typings specified in compiler options
+                if (options.types) {
+                    for (const moduleName of options.types){
```
nit: space between `){`

rbuckton commented on src/services/services.ts:
```diff
@@ -4464,6 +4508,475 @@ namespace ts {
+                            addPotentialPackageNames(package.dependencies, foundModuleNames);
+                        }
+                        if (package.devDependencies) {
+                            addPotentialPackageNames(package.devDependencies, foundModuleNames);
```
do we care about `optionalDependencies` or `peerDependencies` as well?

riknoll replied:
Yeah, I think both of those make sense. Not sure why I left them out originally 😆

rbuckton commented on src/services/services.ts:
```diff
@@ -4464,6 +4508,475 @@ namespace ts {
+            }
+
+            // Returns true if the path is explicitly relative to the script (i.e. relative to . or ..)
+            function isPathRelativeToScript(path: string) {
```
Would `isExternalModuleNameRelative` (in utilities.ts) work in place of this function?

riknoll replied:
Mine is different because it requires the `'.'` characters to be followed by a `'/'`. I could redo mine as a regex though.

rbuckton commented:
:+1: after addressing the remaining comments.

mhegazy commented on src/server/protocol.d.ts:
```diff
@@ -718,6 +718,16 @@ declare namespace ts.server.protocol {
+      * specifiers or paths given that position refers to a module
+      * import declaration, require call, or triple slash reference.
+      */
+    export interface ImportModuleCompletionsRequest extends FileLocationRequest {
```
why is this a separate command?

mhegazy commented on src/services/services.ts:
```diff
@@ -4464,6 +4508,475 @@ namespace ts {
+                        // Enumerate the available files if possible
+                        const files = host.readDirectory(baseDirectory, extensions, /*exclude*/undefined, /*include*/["./*"]);
+                        for (let filePath of files) {
+                            filePath = normalizePath(filePath);
```
if the paths come absolute consider using `normalizeSlashes` instead here to avoid the garbage created by `normalizePath`.

mhegazy commented on src/services/services.ts:
```diff
@@ -4464,6 +4508,475 @@ namespace ts {
+                            }
+
+                            const fileName = includeExtensions ? getBaseFileName(filePath) : removeFileExtension(getBaseFileName(filePath));
+                            const duplicate = !includeExtensions && forEach(result, entry => entry.name === fileName);
```
would it be better to have result be a map, and then convert it to an array at the end to avoid the n^2 searches.

mhegazy commented on src/services/services.ts:
```diff
@@ -4464,6 +4510,477 @@ namespace ts {
+                    }
+                }
+                else if (host.getDirectories && options.typeRoots) {
+                    const absoluteRoots = map(options.typeRoots, rootDirectory => {
```
consider folding this into the loop below and avoid the array allocation.

mhegazy commented on src/services/services.ts:
```diff
@@ -4464,6 +4510,477 @@ namespace ts {
+                        result.push(createCompletionEntryForModule(moduleName, ScriptElementKind.externalModuleName));
+                    }
+                }
+                else if (host.getDirectories && options.typeRoots) {
```
get the effective typeRoots here, see `getEffectiveTypeRoots` in program. we can export this function and use it here as well.

mhegazy commented:
no more comments on my side. i would like to get @dbaeumer's feedback on the design approach though.

mhegazy commented on src/server/session.ts:
```diff
@@ -778,7 +778,17 @@ namespace ts.server {
-                    result.push(entry);
+                    const { name, kind, kindModifiers, sortText, replacementSpan } = entry;
+
+                    let convertedSpan: protocol.TextSpan = undefined;
```
can we limit the creation of the new object to only if the replacement span exists. this avoids creating garbage in the common case (non-module name completion)

mhegazy commented on src/services/services.ts:
```diff
@@ -1475,6 +1488,7 @@ namespace ts {
         kind: string;            // see ScriptElementKind
         kindModifiers: string;   // see ScriptElementKindModifier, comma separated
         sortText: string;
+        replacementSpan?: TextSpan;
```
Please add a comment on when this is set and what to do if it is not.

mhegazy commented:
👍 after the 2 remaining comments

mhegazy commented:
we will need to merge this into release-2.0.5 as well as master.

mhegazy commented:
thanks. 👍

fxck commented:
So is this in 1.5?

mhegazy commented:
VSCode 1.5 ships with TS 1.8 in box. this is likely going to be in the October release of VSCode.

jbrownson commented:
I had been using Atom for TS dev for awhile, experimented w/ Code, seems better in every way except this, it's so nice to have this back and better than it was in Atom. Thanks!

JohannesRudolph commented:
I don't quite get whether this is in VSCode 1.6 (which ships with TS 2.0.3)? I would assume that they'd mentioned such a massive improvement in their release announcements... or did this PR only add the "backend" support to the TS compiler?

fxck commented:
I don't think it's in..
